### PR TITLE
[marketplace] Add subscribe-path 429, env var, and tests for the spend cap

### PR DESCRIPTION
**Pull request by onder-akkaya**
## Summary

Follow-up to `9c29f53` (first commit on this branch), which added `spend_cap.py` (rolling 24h per-wallet USDC spend tracker on a Redis sorted set) and wired the charge-time refusal into `MarketService._charge_one`. That commit flagged three gaps: a subscribe-path HTTP response, an `.env.example` entry, and tests. This PR closes them.

Background on #713: the original spec (RevenueSplit.sol + x402 middleware) never got built as written — PR #958 shipped a subscription marketplace instead. Mapping the old acceptance criteria against main, the one real gap left was a per-user spend cap. Ricardo confirmed a low-cap-with-renew model works for the current payment flow. #713 stays open pending review here.

## What's in it

- `POST /api/marketplace/subscribe` returns 429 for a wallet at/over its rolling 24h cap. Sits after the existing dedup checks and before wallet provisioning. The charge-time refusal in `_charge_one` stays the operationally load-bearing guard (it runs in the tick loop and can't produce an HTTP response itself).
- **Reviewer heads-up: `_charge_one` changed return type**, `bool` → `tuple[bool, str | None]`, to carry the spend-cap halt reason. 11 pre-existing tests in `test_per_step_charging` / `test_service_tick` / `test_settlement_idempotency` mocked or asserted the old bool and are fixed here.
- `.env.example`: `MARKETPLACE_SPEND_CAP_USDC=50`. The default is a placeholder, not a risk-modeled number; 0 disables the check.
- New `backend/tests/marketplace/test_spend_cap.py`: 25 tests, 100% line coverage on `spend_cap.py`, fakeredis-backed (real Lua execution, not mocks).
- Docstring fix on `get_24h_spend_usdc`: it doesn't itself catch Redis errors, `is_over_cap` is the fail-open boundary. No logic change.

## Atomicity fix (from Dan's review)

The original check-then-record pairing (`is_over_cap` read, `record_charge_usdc` write, async `payments.charge()` in between) was a TOCTOU race: N concurrent charges near the cap could all read "under cap" before any of them recorded anything. Replaced with one atomic Lua script (`_CHECK_AND_RESERVE_LUA`, modeled on `redis_state.py`'s `_LEASE_ACQUIRE_LUA`) that prunes the window, sums it, and reserves in a single EVAL. Both call sites (`is_over_cap`, `try_reserve_usdc`) share it.

`_charge_one` reserves immediately before `payments.charge()`, gated behind the settlement-intent idempotency claim so a crash-retry of an already-settled charge short-circuits before ever reaching the reservation. Cost: a cap-refused charge now claims an intent slot (finalized `failed`) where it used to claim none. `record_charge_usdc` is gone. A concurrency test fires 25 concurrent reservations at a 10-USDC cap against real fakeredis Lua and confirms exactly as many succeed as fit, never more.

Known limitation, deliberately not scope-crept in: the 429 doesn't surface a reset time. With a rolling window the honest answer is "when your oldest in-window charge ages out" — happy to add as a follow-up if it's worth the extra Redis read.

## Post-review hardening (`52063f7`)

- `charge_id` now includes `sub_id`. The old `{tick_id}:{step}` form was collision-free only because the subscribe route allows one running sub per (wallet, strategy) — an invariant enforced in a different file. If it ever relaxes, two same-wallet subs in one tick+step would share a Redis member, undercount, and cross-release. One f-string term removes the coupling.
- A raise escaping `payments.charge` (documented "never raises", but the reservation shouldn't depend on that holding forever) used to strand the reserved amount in the window for 24h. Now caught and treated as a failed charge: released, finalized `failed`. Test added for the raise path.

## Verify

```
pytest --cov=archimedes.marketplace.spend_cap --cov-report=term-missing backend/tests/marketplace/test_spend_cap.py
# 25 passed, 100% coverage
env -i HOME=$HOME PATH=$PATH PYTHONPATH=backend python -m pytest backend/tests/ -q -k "not integration"
# 2957 passed, 2 skipped, 0 failed (re-run after 52063f7)
```

ruff format + critical rules (E9,F63,F7,F40) clean on every touched file.

## Review

Needs Dan's review — fund-adjacent charge path. `PAYMENTS_DRY_RUN=true` in prod today, so no real fund movement yet, but this code is live the moment that flag flips. Bogdan preferred second reviewer. Ricardo (`rcrdoh`) owns the marketplace lane per the team table, flagging for his read too.

No release marker (patch): a safety guard on an already-shipped feature, not new user-facing capability. Skipped `arc-canteen update-product` for the same reason.

### CODE REVIEW

**Comment by onder-akkaya** [comment]
Checked the orphan-fn jump (14→129) before treating it as a real signal — it's the same false-positive pattern PR #1076 hit (11→84): the analyzer counts standalone functions with no traceable in-repo caller as orphans, and this PR adds ~30 new functions across `spend_cap.py` and its 25-test companion file, most of which are pytest test functions discovered by pytest's collection mechanism rather than called from other code — invisible to a static call-graph analyzer. CC actually improved (3.0→2.6 avg), and this gate is informational-only per this repo's CI conventions regardless. Not something to fix.

**Comment by onder-akkaya** [comment]
Traced the "85 errors" before assuming it applied to this PR. It's the whole-repo baseline under the full (non-blocking) ruff ruleset — confirmed by running `ruff check` on this branch vs. `main` for the exact 6 pre-existing files this PR touches: 13 errors here vs. 14 on `main` for those same files (net improvement, not a regression), and the 2 brand-new files (`spend_cap.py`, `test_spend_cap.py`) come back completely clean. The hard-blocking gate (`ruff format --check` + the narrow `E9,F63,F7,F40` selector) already passes per the CI run above. Nothing in this diff needs a lint fix.

**Comment by onder-akkaya** [comment]
This PR doesn't touch `requirements.txt` or `environment.yml` at all (`spend_cap.py` only imports stdlib plus the already-present `archimedes.services.redis_state`) — a diff that adds no dependencies can't introduce a new dependency CVE, so the 2 pip-audit findings are pre-existing repo baseline, unrelated to this branch. Nothing to fix here either.

**Comment by dbrowneup** [comment]
Review feedback while this is still draft (full review — diff + spend_cap.py + service.py + the 25 new tests):

**One blocker-class finding: the cap check has a TOCTOU race.** `is_over_cap()` (ZRANGE-sum) and `record_charge_usdc()` (ZADD) are two separate Redis round-trips with the async `payments.charge()` in between (`_charge_one`). The module's own docstring anticipates one wallet running several subscriptions — when N of them tick concurrently near the cap, all N read pass-the-cap simultaneously, all N charge, and the wallet blows through the cap by up to N× a single charge. Same shape on the new subscribe-path 429: check-then-act without atomicity.

**Suggested fix, with an in-repo precedent:** make check-and-reserve a single atomic step — a Lua `EVAL` that sums the window and conditionally ZADDs the reservation in one call, exactly the pattern `redis_state.py`'s `_LEASE_ACQUIRE_LUA` already uses for the runner exactly-once lease. Reserve before `payments.charge()`, remove the reservation on charge failure. That keeps the rolling-window design intact and closes both the charge path and the subscribe path with one primitive.

Minor, take-or-leave: #713's original ask had the 429 body explain "the next reset" — the rolling window makes that ambiguous (oldest entry's expiry is the honest answer), and Ricardo's simplified cap model may have rescoped it; if so, one sentence in the PR body saying it was dropped intentionally would close the loop.

Everything else reads well — the env-var wiring with `.env.example` documentation, the 429 semantics, and the test coverage over the pre-existing `_charge_one` signature change are all solid. Happy to re-review when it leaves draft.

**Comment by onder-akkaya** [comment]
Fixed the race with the atomic Lua check-and-reserve you described: `spend_cap._CHECK_AND_RESERVE_LUA`, modeled on `_LEASE_ACQUIRE_LUA`. `is_over_cap` and the new `try_reserve_usdc` both run through it now, so the subscribe path and the charge path share one primitive.

One deliberate deviation from "reserve right where the old check was": the reservation now sits right before `payments.charge()`, gated behind the settlement-intent idempotency claim instead of ahead of it. Reserving ahead of the claim double-counts a crash-retry of an already-settled charge, since that charge_id's entry is already in the window from the first successful reservation and would get summed again on the retry. The claim's uniqueness is what makes the reservation safe to treat as "runs at most once." Cost is a `failed` SettlementIntent row on a cap refusal where there used to be none. Seemed worth it to close the double-count edge case cleanly instead of special-casing it in the Lua script.

Added a concurrency test that fires 25 concurrent reservations at a pinned 10-USDC cap through real fakeredis+Lua execution, not mocked, and confirms exactly as many succeed as the cap allows.

Re: the reset wording, you're right the rolling window makes a single instant ambiguous. Dropped it intentionally rather than rescoped by Ricardo. Added a sentence in the PR body on what the honest answer would look like (oldest in-window entry's expiry) if it's worth a follow-up.

Also installed circlekit locally since I was in here anyway (the exact pin already in environment.yml). Full backend suite now actually runs: 2910 passed, 4 skipped, 0 failed.

File: backend/archimedes/marketplace/spend_cap.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    local amt = tonumber(string.match(m, ".*:(%d+)$"))
+    if amt then
+        total = total + amt
+    end
+end
+
+local projected = total
+if already_reserved == false then
+    projected = total + amount_raw
+end
```
[the remark this answers is not in the document]
Comment (onder-akkaya) [reply backend/archimedes/marketplace/spend_cap.py:111]: Confirmed and fixed. Reproduced it standalone first: seeded a real "" member against the old script and it reported "fits under cap" for a hypothetical amount exactly equal to the cap, which should have been refused. Gated the ZSCORE existence check behind do_reserve==1 and defaulted the projection to total + amount_raw, so the read-only path (is_over_cap) never runs it and never has this failure mode. Added a regression test pinning the exact scenario (commit a34c3f9).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by danielscoffee — APPROVED** [review]
LGTM. But important warning: quality-gate might have conflict with PR: #1152

**Comment by dbrowneup** [comment]
Merged — the cap is a clear improvement over no cap, and the honesty about the default being a placeholder rather than a risk-modelled number is right.

One thing I want on the record rather than passing unremarked, because it is the failure shape this repo has been working through all week:

\`\`\`python
raw = os.getenv("MARKETPLACE_SPEND_CAP_USDC", "50")
try:
    return Decimal(raw)
except Exception:
    logger.warning("... is not a valid number — treating as disabled")
    return Decimal(0)
\`\`\`

A **typo** in the env var silently disables the spend cap entirely. `MARKETPLACE_SPEND_CAP_USDC="50 USDC"`, or a stray quote from a shell heredoc, and the guard is gone — with a WARNING nobody reads. This is the same shape as `AGENT_DRY_RUN=1` silently meaning LIVE (caught on #1173/#1174), and as the oracle logging `"Prices fetched"` every 60s for three weeks while never pushing a price.

The unset case is fine — the default is `"50"`, so unset means enabled. It is specifically the **malformed** case that fails open.

Two-line fix: fall back to the DEFAULT rather than to disabled, and reserve `0` for an explicit, deliberate opt-out:

\`\`\`python
except Exception:
    logger.error("MARKETPLACE_SPEND_CAP_USDC=%r is not a valid number — falling back to the default cap, NOT disabling", raw)
    return Decimal("50")
\`\`\`

Not blocking the merge: marketplace is out of MVP scope, `PAYMENTS_DRY_RUN=true`, and a cap that can be misconfigured still beats no cap at all. But the docstring frames fail-open-when-unconfigured as "this codebase's general convention", and that convention is precisely what we have spent the week dismantling — it should not be cited as precedent on a funds path. I will pick this up in a follow-up unless you would rather take it.
